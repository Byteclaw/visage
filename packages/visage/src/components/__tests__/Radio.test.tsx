/* eslint-disable no-irregular-whitespace */
import { fireEvent } from '@testing-library/react';
import React from 'react';
import { render, createTestTheme } from './render';
import { Radio } from '../Radio';

describe('Radio', () => {
  it('respects disabled prop', () => {
    const handler = jest.fn();
    const { getByTestId } = render(
      <Radio
        data-testid="radio"
        disabled
        onClick={handler}
        onKeyDown={handler}
        label=""
        labelProps={{ 'data-testid': 'label' }}
      />,
    );

    expect(getByTestId('radio')).toHaveAttribute('disabled');
    expect(getByTestId('label')).toHaveAttribute('data-disabled', 'true');

    fireEvent.click(getByTestId('radio'));
    // we don't fire a keyDown because it isn't possible to focus disabled checkbox in browser

    expect(handler).not.toHaveBeenCalled();
  });

  it('respects hiddenLabel prop', () => {
    const { getByTestId } = render(
      <Radio
        data-testid="radio"
        hiddenLabel
        label="Label"
        labelProps={{ 'data-testid': 'label' }}
        labelTextProps={{ 'data-testid': 'label-text' }}
      />,
    );

    expect(getByTestId('label-text')).toHaveAttribute('data-hidden', 'true');
  });

  it('respects invalid prop', () => {
    const props = {
      'data-testid': 'radio',
      label: 'Label',
      labelProps: { 'data-test-id': 'label' },
      labelTextProps: { 'date-testid': 'label-text' },
    };
    const { getByTestId, rerender } = render(<Radio {...props} />);

    expect(getByTestId('radio')).not.toHaveAttribute('aria-invalid');

    rerender(<Radio {...props} invalid />);

    expect(getByTestId('radio')).toHaveAttribute('aria-invalid', 'true');
  });

  it('respects readOnly prop', () => {
    const handler = jest.fn();
    const { getByTestId } = render(
      <Radio
        data-testid="radio"
        readOnly
        onClick={handler}
        onKeyDown={handler}
        label=""
        labelProps={{ 'data-testid': 'label' }}
      />,
    );

    // read only checkbox can be focused, clicked, etc but can't be changed
    expect(getByTestId('radio')).toHaveAttribute('readOnly');
    expect(getByTestId('label')).toHaveAttribute('data-readonly', 'true');

    fireEvent.click(getByTestId('radio'));
    fireEvent.keyDown(getByTestId('radio'), { key: ' ' });

    expect(getByTestId('radio')).not.toHaveAttribute('checked');

    expect(handler).toHaveBeenCalledTimes(2);
  });

  it('renders correctly', () => {
    const { asFragment } = render(<Radio label="Required label" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('allows to customize Toggler (uncontrolled)', () => {
    const toggler = jest.fn(() => <div />);
    const { container, rerender } = render(<Radio invalid toggler={toggler} />);

    expect(toggler).toHaveBeenCalledTimes(1);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: false,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    // now change the value
    fireEvent.click(container.querySelector('input[type=radio]'));

    expect(toggler).toHaveBeenCalledTimes(2);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: true,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    fireEvent.focus(container.querySelector('input[type=radio]'));

    expect(toggler).toHaveBeenCalledTimes(3);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: true,
        disabled: undefined,
        focused: true,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    fireEvent.blur(container.querySelector('input[type=radio]'));

    expect(toggler).toHaveBeenCalledTimes(4);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: true,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    // make it readonly
    rerender(
      <Radio
        defaultChecked={false}
        key="1"
        invalid
        readOnly
        toggler={toggler}
      />,
    );

    expect(toggler).toHaveBeenCalledTimes(5);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: false,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: true,
      },
      expect.anything(),
    );

    // toggle
    fireEvent.click(container.querySelector('input[type=radio]'));

    expect(toggler).toHaveBeenCalledTimes(5);
  });

  it('allows to customize Toggler (controlled)', () => {
    const toggler = jest.fn(() => <div />);
    const onChange = jest.fn();
    const { container, rerender } = render(
      <Radio checked={false} invalid onChange={onChange} toggler={toggler} />,
    );

    expect(toggler).toHaveBeenCalledTimes(1);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: false,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    // now change the value
    fireEvent.click(container.querySelector('input[type=radio]'));

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(toggler).toHaveBeenCalledTimes(1);

    // rerender as checked
    rerender(<Radio checked invalid onChange={onChange} toggler={toggler} />);

    expect(toggler).toHaveBeenCalledTimes(2);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: true,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    fireEvent.focus(container.querySelector('input[type=radio]'));

    expect(toggler).toHaveBeenCalledTimes(3);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: true,
        disabled: undefined,
        focused: true,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );

    fireEvent.blur(container.querySelector('input[type=radio]'));

    expect(toggler).toHaveBeenCalledTimes(4);
    expect(toggler).toHaveBeenLastCalledWith(
      {
        checked: true,
        disabled: undefined,
        focused: false,
        invalid: true,
        readOnly: undefined,
      },
      expect.anything(),
    );
  });

  it('is extendable by styles', () => {
    const { asFragment } = render(
      <Radio label="test" styles={{ color: 'pink' }} />,
    );

    // label should have color pink
    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by RadioLabel face', () => {
    const { asFragment } = render(<Radio label="test" />, {
      ds: {
        theme: createTestTheme({ RadioLabel: { borderRadius: 10 } }),
      },
    });

    // emotion 4 should have border-radius: 10px;
    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by RadioLabelText face', () => {
    const { asFragment } = render(<Radio label="test" />, {
      ds: {
        theme: createTestTheme({ RadioLabelText: { borderRadius: 10 } }),
      },
    });

    // emotion 3 should have border-radius: 10px;
    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by RadioControl face', () => {
    const { asFragment } = render(<Radio label="test" />, {
      ds: {
        theme: createTestTheme({ RadioControl: { borderRadius: 10 } }),
      },
    });

    // emotion 0 should have border-radius: 10px;
    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by RadioToggler face', () => {
    const { asFragment } = render(<Radio label="test" />, {
      ds: {
        theme: createTestTheme({ RadioToggler: { borderRadius: 10 } }),
      },
    });

    // emotion 2 should have border-radius: 10px;
    expect(asFragment()).toMatchSnapshot();
  });
});
