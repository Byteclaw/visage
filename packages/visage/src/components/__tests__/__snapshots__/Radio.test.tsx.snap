// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio is extendable by RadioControl face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4[data-disabled="true"] {
  cursor: default;
  pointer-events: none;
  resize: none;
  opacity: 0.5;
}

.emotion-4[data-readonly="true"] {
  cursor: default;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-radius: 999px;
  border-style: solid;
  border-width: 2px;
  border-color: rgb(0,0,255);
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  fill: rgb(209,209,209);
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-3[data-hidden="true"] {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  border-radius: 10px;
}

<label
    class="emotion-4"
    data-disabled="false"
    data-readonly="false"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="6"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio is extendable by RadioLabel face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-radius: 999px;
  border-style: solid;
  border-width: 2px;
  border-color: rgb(0,0,255);
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  fill: rgb(209,209,209);
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-3[data-hidden="true"] {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 10px;
}

.emotion-4[data-disabled="true"] {
  cursor: default;
  pointer-events: none;
  resize: none;
  opacity: 0.5;
}

.emotion-4[data-readonly="true"] {
  cursor: default;
}

<label
    class="emotion-4"
    data-disabled="false"
    data-readonly="false"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="6"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio is extendable by RadioLabelText face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4[data-disabled="true"] {
  cursor: default;
  pointer-events: none;
  resize: none;
  opacity: 0.5;
}

.emotion-4[data-readonly="true"] {
  cursor: default;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-radius: 999px;
  border-style: solid;
  border-width: 2px;
  border-color: rgb(0,0,255);
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  fill: rgb(209,209,209);
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
  border-radius: 10px;
}

.emotion-3[data-hidden="true"] {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<label
    class="emotion-4"
    data-disabled="false"
    data-readonly="false"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="6"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio is extendable by RadioToggler face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4[data-disabled="true"] {
  cursor: default;
  pointer-events: none;
  resize: none;
  opacity: 0.5;
}

.emotion-4[data-readonly="true"] {
  cursor: default;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  width: 1em;
  height: 1em;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-3[data-hidden="true"] {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-radius: 10px;
  border-style: solid;
  border-width: 2px;
  border-color: rgb(0,0,255);
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  fill: rgb(209,209,209);
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

<label
    class="emotion-4"
    data-disabled="false"
    data-readonly="false"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="6"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio is extendable by styles 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-radius: 999px;
  border-style: solid;
  border-width: 2px;
  border-color: rgb(0,0,255);
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  fill: rgb(209,209,209);
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-3[data-hidden="true"] {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: rgb(255,192,203);
}

.emotion-4[data-disabled="true"] {
  cursor: default;
  pointer-events: none;
  resize: none;
  opacity: 0.5;
}

.emotion-4[data-readonly="true"] {
  cursor: default;
}

<label
    class="emotion-4"
    data-disabled="false"
    data-readonly="false"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="6"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      test
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-4 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  line-height: inherit;
  margin: 0px;
  outline: none;
  padding: 0px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4[data-disabled="true"] {
  cursor: default;
  pointer-events: none;
  resize: none;
  opacity: 0.5;
}

.emotion-4[data-readonly="true"] {
  cursor: default;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: rgb(209,209,209);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  border-radius: 999px;
  border-style: solid;
  border-width: 2px;
  border-color: rgb(0,0,255);
  margin-right: 8px;
}

.emotion-2[data-checked="true"] {
  background-color: light-blue;
  color: rgb(255,255,255);
  border-color: light-blue;
}

.emotion-2[data-checked="true"] > svg {
  fill: rgb(209,209,209);
  visibility: visible;
}

.emotion-2[data-checked="false"] > svg {
  visibility: hidden;
}

.emotion-2[data-invalid="true"] {
  border-color: rgb(255,0,0);
}

.emotion-2[data-invalid="true"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(255,0,0);
}

.emotion-2[data-invalid="false"][data-focused="true"] {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-1 {
  width: 1em;
  height: 1em;
}

.emotion-3 {
  font-size: inherit;
  line-height: inherit;
}

.emotion-3[data-hidden="true"] {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<label
    class="emotion-4"
    data-disabled="false"
    data-readonly="false"
  >
    <input
      class="emotion-0"
      type="radio"
    />
    <div
      aria-hidden="true"
      class="emotion-2"
      data-checked="false"
      data-disabled="false"
      data-focused="false"
      data-invalid="false"
      data-readonly="false"
    >
      <svg
        class="emotion-1"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="6"
        />
      </svg>
    </div>
    ​ 
    <span
      class="emotion-3"
      data-hidden="false"
    >
      Required label
    </span>
  </label>
</DocumentFragment>
`;
